rfcs/2021-10-29-8621-framing-and-codecs-sinks.md

```diff
@@ -0,0 +1,126 @@
+
+```rust
+/// An encoder that can encode structured events to byte messages.
+pub struct Encoder {
+    serializer: BoxedSerializer,
+    framer: BoxedFramer,
+}
+```
+
+`Encoder` implements `tokio_util::codec::Encoder<SmallVec<[Event; 1]>>`. Internally, events first go through the `Serializer` which implements `tokio_util::codec::Encoder<SmallVec<[Event; 1]>>` and are then handed over to the `Framer` which implements `tokio_util::codec::Encoder<Bytes>`.
+
+## Plan Of Attack
+
+Incremental steps to execute this change. These will be converted to issues after the RFC is approved:
+
+- [ ] Implementation of support structures (`Encoder`, `SerializerConfig`, `FramingConfig`)
+- [ ] Implementation of selected `encoders`/`framers` (e.g. `JSON` and `newline_delimited`)
+- [ ] Example integration with first sink, e.g. `socket` or `http` which benefit most from generic encoding/framing options
+- [ ] Subsequent PRs for each integration to a sink
+
+## Surveyed Sinks
+
+Overview for the current state of sinks regarding encoding:
+
+|sink|encoding config|notes|
+|-|-|-|
+|`aws_cloudwatch_logs`| `EncodingConfig<Encoding { Text, Json }>` | Enveloped in `rusoto_logs::InputLogEvent`. `Text` reads message_key()
+|`aws_kinesis_firehose`| `EncodingConfig<Encoding { Text, Json }>` | Enveloped in `rusoto_firehose::Record` that serializes to base64. `Text` reads `message_key()`
+|`aws_kinesis_streams`| `EncodingConfig<Encoding { Text, Json }>` | Enveloped in `rusoto_kinesis::PutRecordsRequestEntry`. `Text` reads `message_key()`
+|`aws_s3`| `EncodingConfig<StandardEncodings { Text, Json, Ndjson }>` | Uses util::{RequestBuilder, Encoder, Compressor}. `Text` reads `message_key()`
+|`aws_sqs`| `EncodingConfig<Encoding { Text, Json }>` | Enveloped in EncodedEvent<SendMessageEntry>. `Text` reads `message_key()`
+|`azure_blob`| `EncodingConfig<Encoding { Ndjson, Text }>` | Enveloped in EncodedEvent<PartitionInnerBuffer>. `Text` reads `message_key()`
+|`azure_monitor_logs`| `EncodingConfigWithDefault<Encoding { Default }>` | Reshapes events only, without encoding
+|`blackhole`| -/-
+|`clickhouse`| `EncodingConfigWithDefault<Encoding { Default }>` | Reshapes events only, without encoding
+|`console`| `EncodingConfig<Encoding { Text, Json }>` | `Text` reads `message_key()`
+|`datadog`| `EncodingConfigFixed<DatadogLogsJsonEncoding>` | Doesn't provide options to encode the event payload separately from the protocol
+|`datadog_archives`|
+|`elasticsearch`|
+|`file`| `EncodingConfig<Encoding { Text, Ndjson }>` | `Text` reads `message_key()`
+|`gcp`|
+|`honeycomb`|
+|`http`| `EncodingConfig<Encoding { Text, Ndjson, Json }>` | Enveloped in HTTP request. Request-level compression. Sets headers depending on encoding config
+|`humio`|
+|`influxdb`|
+|`kafka`| `EncodingConfig<StandardEncodings { Text, Json, Ndjson }>` | Doesn't reshape, uses `Encoder<Event>` for `StandardEncodings` in `encode_input`, enveloped in `KafkaRequest`
+|`logdna`|
+|`loki`|
+|`nats`|
```

Comment:
The `nats` sink supports encoding, doesn't it?

https://github.com/vectordotdev/vector/blob/d3a4492ee7f250c30c6e26592a7d8d4fdb3de70f/src/sinks/nats.rs#L194-L205